File: app/src/main/java/com/batodev/arrows/engine/ExtraParameterObjects.kt
package com.batodev.arrows.engine

import androidx.compose.ui.graphics.Color

data class SnakeRecursiveParams(
    val config: GameGeneratorConfig,
    val occupied: Array<BooleanArray>,
    val snakes: List<Snake>,
    val body: List<Point>,
    val forbidden: Set<Point>,
    val criterion: Criterion,
    val prevDir: Direction?
)

data class TapParams(
    val snake: Snake,
    val isVibrationEnabled: Boolean,
    val obstructed: Boolean,
    val lives: Int,
    val onPenalty: () -> Unit,
    val onSuccess: () -> Unit
)

data class DebugDialogParams(
    val dialogToShow: String?,
    val viewModel: com.batodev.arrows.ui.AppViewModel,
    val levelNumber: Int,
    val forcedWidth: Int?,
    val forcedHeight: Int?,
    val forcedLives: Int?,
    val forcedShape: String?,
    val onDismiss: () -> Unit
)

File: flutter_arrows/lib/v3/modules/game/application/models/extra_parameter_objects.dart
import 'board_point.dart';
import 'direction.dart';
import 'game_generator_models.dart';
import 'generation_criterion.dart';
import 'snake.dart';

class SnakeRecursiveParams {
  final GameGeneratorConfig config;
  final List<List<bool>> occupied;
  final List<Snake> snakes;
  final List<BoardPoint> body;
  final Set<BoardPoint> forbidden;
  final Criterion criterion;
  final Direction? prevDir;

  const SnakeRecursiveParams({
    required this.config,
    required this.occupied,
    required this.snakes,
    required this.body,
    required this.forbidden,
    required this.criterion,
    this.prevDir,
  });

  SnakeRecursiveParams copyWith({
    List<BoardPoint>? body,
    Direction? prevDir,
  }) {
    return SnakeRecursiveParams(
      config: config,
      occupied: occupied,
      snakes: snakes,
      body: body ?? this.body,
      forbidden: forbidden,
      criterion: criterion,
      prevDir: prevDir ?? this.prevDir,
    );
  }
}
